### PR TITLE
Use dbWriteTable Copy in dbAppendTable to Eliminate Slowdown

### PR DESCRIPTION
**Pull request by JSchoenbachler**
This is a pull request to close https://github.com/r-dbi/RPostgres/issues/241 

The only change is to replace the dbAppendTable function with the relevant pieces of dbWriteTable when append & copy = TRUE.

### CODE REVIEW

File: R/tables.R [hunk 1]
```diff
@@ -191,17 +191,19 @@ setMethod("dbAppendTable", c("PqConnection"),
...
+    value <- sql_data_copy(value, row.names = FALSE)
+
+    fields <- dbQuoteIdentifier(conn, names(value))
+    sql <- paste0(
+      "COPY ", dbQuoteIdentifier(conn, name),
+      " (", paste(fields, collapse = ", "), ")",
+      " FROM STDIN"
+    )
+    connection_copy_data(conn@ptr, sql, value)
   }
```
Comment (krlmlr) [inline R/tables.R:207]: Need to return a value here. Test with `DBItest::test_some("append_table_return")`. Also, factors should give a warning: `DBItest::test_some("append_roundtrip_factor")`

**Review by krlmlr — COMMENTED** [review]
Thanks for working on this!

File: R/tables.R (hunk 1 above)
Comment (JSchoenbachler) [reply R/tables.R]: @krlmlr I have it now returning `nrow(values)` and a warning if factors are supplied. Let me know if there is anything else I need to do!

File: R/tables.R [hunk 2]
```diff
@@ -191,17 +191,26 @@ setMethod("dbAppendTable", c("PqConnection"),
...
-      row.names = row.names,
-      prefix = "$",
-      pattern = "1",
-      ...
+    row.names <- FALSE
+
+    is_factor = vlapply(value, is.factor)
+    if (any(is_factor)) {
+      warning("Factors converted to character", call. = FALSE)
+    }
```
Comment (krlmlr) [inline R/tables.R:199]: Does this work:

```suggestion
    value = factor_to_string(value)
```

**Review by krlmlr — COMMENTED** [review]
Thanks, this looks great, tests pass. Can we reuse code?

File: R/tables.R (hunk 2 above)
Comment (JSchoenbachler) [reply R/tables.R:199]: This is fine with me!

**Comment by JSchoenbachler** [comment]
@krlmlr Approved suggestion and committed to branch!

File: R/tables.R [hunk 3]
```diff
@@ -191,17 +191,23 @@ setMethod("dbAppendTable", c("PqConnection"),
...
-      con = conn,
-      table = name,
-      values = value,
-      row.names = row.names,
-      prefix = "$",
-      pattern = "1",
-      ...
+    row.names <- FALSE
+
+    value = factor_to_string(value)
```
Comment (krlmlr) [inline R/tables.R:196]:
```suggestion
    value = factor_to_string(value, warn = TRUE)
```

**Review by krlmlr — COMMENTED** [review]
Looks like we need `warn = TRUE`.

File: R/tables.R (hunk 3 above)
Comment (JSchoenbachler) [reply R/tables.R:196]: Yep, that works!

**Comment by JSchoenbachler** [comment]
Commit is done!

**Comment by krlmlr** [comment]
Thanks! I think `dbAppendTable()` needs a `copy = TRUE` argument, similarly to `dbWriteTable()`, because `COPY` doesn't work everywhere.

I should have noticed earlier. Would you like to include this in this PR? If not, no worries, I can take this on.

Also, `dbWriteTable()` ultimately should forward to `dbAppendTable()` and also `dbCreateTable()`. We could do that in a separate PR.

**Comment by JSchoenbachler** [comment]
@krlmlr Could you review the changes I made in the most recent commit and let me know if that is the way you want those things addressed? The only thing I think I didn't add was the addition of using `dbCreateTable()`.

**Comment by krlmlr** [comment]
Thanks. Could you please run `devtools::document()`? Please watch out for failures in the automated tests.

**Comment by JSchoenbachler** [comment]
@krlmlr All checks passed. Had to add a param for warn. I will add the `dbCreateTable` function call.

**Comment by JSchoenbachler** [comment]
@krlmlr wait, I just checked and it already looks like `dbWriteTable()` calls `dbCreateTable()` on line 103 of tables.R. Was I misunderstanding what you were asking?

**Comment by JSchoenbachler** [comment]
@krlmlr Hey just wanted to make sure you had seen the above message asking for clarification!

File: R/tables.R [hunk 4]
```diff
@@ -188,20 +174,33 @@ format_keep_na <- function(x, ...) {
 #' @rdname postgres-tables
 #' @export
 setMethod("dbAppendTable", c("PqConnection"),
-  function(conn, name, value, ..., row.names = NULL) {
+  function(conn, name, value, copy = TRUE, warn = TRUE, ..., row.names = NULL) {
```
Comment (krlmlr) [inline R/tables.R:177]: We don't need a `warn` argument here, `dbAppendTable()` is a new generic.

```suggestion
  function(conn, name, value, ..., copy = TRUE, row.names = NULL) {
```

File: R/tables.R [hunk 5]
```diff
@@ -188,20 +174,33 @@ format_keep_na <- function(x, ...) {
...
-      table = name,
-      values = value,
-      row.names = row.names,
-      prefix = "$",
-      pattern = "1",
-      ...
-    )
+    row.names <- FALSE
+
+    value = factor_to_string(value, warn = warn)
```
Comment (krlmlr) [inline R/tables.R:182]:
```suggestion
    value = factor_to_string(value, warn = TRUE)
```

File: R/tables.R [hunk 6]
```diff
@@ -188,20 +174,33 @@ format_keep_na <- function(x, ...) {
...
+
+    value <- sqlRownamesToColumn(value, row.names)
+
+    if (!copy) {
+      value <- sqlData(conn, value, row.names = FALSE)
+
+      sql <- sqlAppendTable(conn, name, value, row.names = FALSE)
```
Comment (krlmlr) [inline R/tables.R:189]: I wonder if the previous approach via `sqlAppendTableTemplate()` works with Redshift? Can we implement a `dbAppendTable()` method for a `RedshiftConnection` that uses `sqlAppendTable()` and keep using `sqlAppendTableTemplate()` in the `PqConnection` method?

File: R/tables.R [hunk 7]
```diff
@@ -188,20 +174,33 @@ format_keep_na <- function(x, ...) {
...
-      pattern = "1",
-      ...
-    )
+    row.names <- FALSE
+
+    value = factor_to_string(value, warn = warn)
+
+    value <- sqlRownamesToColumn(value, row.names)
+
+    if (!copy) {
```
Comment (krlmlr) [inline R/tables.R:186]: What happens here if the input has zero rows? Do we have a test for this in DBItest?

File: R/tables.R [hunk 8]
```diff
@@ -108,22 +109,7 @@ setMethod("dbWriteTable", c("PqConnection", "character", "data.frame"),
     }
 
     if (nrow(value) > 0) {
```
Comment (krlmlr) [inline R/tables.R:111]: We shouldn't need to check here.

**Review by krlmlr — COMMENTED** [review]
Thanks, I was mistaken, `dbWriteTable()` seems fine. Added a few more comments, would you like to do another round?

**Comment by krlmlr** [comment]
@JSchoenbachler: We are in the process of relicensing RPostgres under MIT, #312. Are you ok with maintaining your contribution under that license?

**Comment by JSchoenbachler** [comment]
@krlmlr yes, that is fine! Sorry I haven't looked at this in a long time.

**Comment by krlmlr** [comment]
No worries! I can pick this up in the next few days, feel free to add more commits in the meantime.

**Pull request closed without merging**